### DISCUSSION TITLE
Allow pattern matching to satisfy type parameter constraints

### DISCUSSION THREAD

**Discussion by the-avid-engineer** [dotnet/csharplang#5252] [Language Ideas]
### Is your feature request related to a problem? Please describe.
It seems like pattern matching and type parameter constraints would work hand-in-hand, but they currently do not (at least as of .NET 5); you cannot use pattern matching to satisfy type parameter constraints.

https://dotnetfiddle.net/batTk7

```
public class Program
{
	public static void Main()
	{
		var foo = new Foo<object>();
		
		DoUnconstrainedThing(foo);
	}
	
	public static void DoUnconstrainedThing<T>(T thing)
	{
		if (thing is IFoo<T> foo) {
			DoConstrainedThing<T>(foo);	
		}
	}
	
	public static void DoConstrainedThing<T>(IFoo<T> foo) where T : IFoo<T> {}
}
	
public interface IFoo<T> {}

public class Foo<T> : IFoo<T> {}
```

This code results in a compiler error; the last sentence of that error is `There is no boxing conversion or type parameter conversion from 'T' to 'IFoo<T>'.`

However, it is established by pattern matching that `T` does in fact implement `IFoo<T>` within the scope of that if block.

### Describe the solution you'd like
It seems to me like this should be a supported way to satisfy a generic constraint

**Comment by jaredpar** [comment]
Moving to csharplang as this is a language design question.

**Comment by HaloFour** [comment]
This seems to want to have the compiler reinterpret what `T` is based on flow analysis of the code.  Wouldn't this fall into the same category of having the types of existing variables reinterpreted?  I'd think that this would make things particularly tricky with invoking static members of the generic type based on the type constraints if the type constraints were fluid as suggested.

I also think that even if the compiler had these "smarts" that the runtime itself would not allow `T` to be used as the generic type argument in that case as `T` is itself not constrained.  You'd need similar flow analysis built into the runtime.

**Comment by the-avid-engineer** [comment]
I cant say I know much about how type constraints (or any features) actually work under the hood. My naïve assumption was that it was just a compile-time feature. If the runtime is checking type constraints, then the benefit of this feature request is definitely outweighed by the cost

**Reply by the-avid-engineer** [reply]
`is definitely outweighed` may be too opinionated.. `may be outweighed` is better.

Revisiting this two years later, it would be interesting if the language could support a different kind of block, leaving `if` as it is. Similar to the `where` clause on methods, make a `when` block that can be used inside a method, but doesn't cover the entire method.

```
public static void DoUnconstrainedThing<T>(T thing)
{
    // this block only executes if T implements IFoo<T>
    when T : IFoo<T>
    {
        // inside this block:
        // - static interface members of IFoo<T> can be accessed, such as T.DoStaticThing()
        // - all instances of T can be used as an instance of IFoo<T>, no pattern matching needed
    }
    
    // Do some other things that aren't constrained to IFoo<T>
}
```

Under the hood / at runtime, this is an `if` block that only executes if the generic parameter constraints of the when block are satisfied. I presume the metadata would be available for checking, in the same way that it's already used at runtime for `where` constraints. The difference would be conditional execution instead of (an exception? not sure what happens if you trick the compiler), and the constraint is for a block inside the method block instead of the entire method block.

**Reply by HaloFour** [reply]
See: #6308

**Reply by the-avid-engineer** [reply]
Yes! I would probably prefer a new keyword, to avoid collisions with existing code, but these are basically the same idea :)

**Reply by the-avid-engineer** [reply]
Then again.. collision wouldn't happen in the code. Maybe no new keyword

**Comment by Joe4evr** [comment]
Prior/related discussions: https://github.com/dotnet/runtime/discussions/58507 https://github.com/dotnet/csharplang/issues/3050 https://github.com/dotnet/csharplang/discussions/662

**Reply by the-avid-engineer** [reply]
Making a new pseudo generic type parameter that can only be used inside the if-block where it is declared via pattern matching does seem like an interesting solution

**Comment by bernd5** [comment]
Isn't the compiler right here? T is `Foo<object>`. It is an `IFoo<object>`. But `object` is not an IFoo...

**Reply by DavidArno** [reply]
That's the point. There's no way to express what the OP is trying to do in C# as it currently stands. Effectively, what's needed is something like:

```c#
public static void DoUnconstrainedThing<T>(T thing)
{
    if (thing is IFoo<_> foo) {
        DoConstrainedThing(foo);	
    }
}
```
There's no way to work with dynamic generics at runtime without using reflection. This - and other - requests are for the language to support such features without having to use reflection.

**Reply by the-avid-engineer** [reply]
I have poorly illustrated what you would pass into DoUnconstrainedThing

A better `Foo` would be:

`public class Foo : IFoo<Foo>`

(No generic parameter on Foo)

It still results in the same compiler error.

**Reply by the-avid-engineer** [reply]
Or if Foo still had a generic parameter, it would be

`public class Foo<T> : IFoo<Foo<T>>`

**Comment by Richiban** [comment]
> However, it is established by pattern matching that T does in fact implement IFoo<T> within the scope of that if block.

That's quite advanced type inference you're talking about there, where pattern matching has not just type-checked the variable `foo` but also deduced that "since `foo` is of type `T` and in this code block `foo` is also `IFoo<T>` then that means that at this line of code `T` must also be assignable to `IFoo<T>`".

C# doesn't currently have _any_ flow-sensitive typing which would be necessary for the generic type parameter `T` to change its meaning from one block of code to another.

**Reply by the-avid-engineer** [reply]
I'm really curious to understand how type constraints currently work under the hood / why they can apply to an entire method block or class/struct block, but not an if block.

**Reply by HaloFour** [reply]
Generic type constraints are part of the metadata of the generic type parameter itself.  The constraints don't have any other scoping, if the generic type parameter is a part of a type declaration then the constraints apply for that entire type.  The runtime enforces that a generic type argument meets the constraints otherwise it will throw an exception.

**Reply by the-avid-engineer** [reply]
Ah, `if` blocks don't have metadata ^_^
